cf-net/cf-net.c

```diff
@@ -0,0 +1,823 @@
+                opts->verbose = true;
+                break;
+            }
+            case 'I':
+            {
+                opts->inform = true;
+                break;
+            }
+            default:
+            {
+                printf("Default optarg = '%s', c = '%c' = %i",
+                       optarg, c, (int)c);
+                break;
+            }
+        }
+    }
+    (*args) = &(argv[optind]);
+    return 0;
+}
+
+// This converts a command string to an enum: "HELP" -> HELP
+static int CFNetCommandNumber(char* command)
+{
+    for (int i = 0; i < command_enum_max; ++i)
+    {
+        if (strcmp(command, command_strings[i]) == 0)
+        {
+            return i;
+        }
+    }
+    return -1;
+}
+
+// ^ Macro returns on match(!)
+static int CFNetCommandSwitch(CFNetOptions *opts, const char *hostname,
+                              char **args, enum command_enum cmd)
+{
+    switch (cmd) {
+        case CONNECT:
+            return CFNetConnect(hostname, args);
+        case STAT:
+            return CFNetStat(opts, hostname, args);
+        case GET:
+            return CFNetGet(opts, hostname, args);
+        case OPENDIR:
+            return CFNetOpenDir(opts, hostname, args);
+        case MULTI:
+            return CFNetMulti(hostname);
+        case MULTITLS:
+            return CFNetMultiTLS(hostname);
+        default:
+            break;
+    }
+    printf("Bug: CFNetCommandSwitch() is unable to pick a command");
+    exit(EXIT_FAILURE);
+    return -1;
+}
+
+static void CFNetSetVerbosity(CFNetOptions *opts)
+{
+    assert(opts);
+    if(opts->debug)
+    {
+        LogSetGlobalLevel(LOG_LEVEL_DEBUG);
+        Log(LOG_LEVEL_DEBUG, "Debug log level enabled");
+    }
+    else if(opts->verbose)
+    {
+        LogSetGlobalLevel(LOG_LEVEL_VERBOSE);
+        Log(LOG_LEVEL_VERBOSE, "Verbose log level enabled");
+    }
+    else if(opts->inform)
+    {
+        LogSetGlobalLevel(LOG_LEVEL_INFO);
+        Log(LOG_LEVEL_INFO, "Inform log level enabled");
+    }
+}
+
+static int CFNetRun(CFNetOptions *opts, char **args, char *hostnames)
+{
+    CFNetSetVerbosity(opts);
+
+    char *command_name = args[0];
+    ToUpperStrInplace(command_name);
+    enum command_enum cmd = CFNetCommandNumber(command_name);
+    if (cmd == HELP)
+    {
+        return CFNetHelp(args[1]);
+    }
+
+    Log(LOG_LEVEL_VERBOSE,
+        "Running command '%s' with argument(s):\n", command_name);
+    for (int i = 1; args[i] != NULL; ++i)
+    {
+        Log(LOG_LEVEL_VERBOSE, "argv[%i]=%s\n", i, args[i]);
+    }
+
+    CFNetInit();
+    char *hosts = RequireHostname(hostnames);
+    int ret = 0;
+    char *hostname = strtok(hosts, ",");
+    if (cmd == -1)
+    {
+        Log(LOG_LEVEL_ERR, "'%s' is not a valid cf-net command\n", args[0]);
+        return -1;
+    }
+    while (hostname != NULL){
+        CFNetCommandSwitch(opts, hostname, args, cmd);
+        hostname = strtok(NULL, ",");
+    }
+    free(hosts);
+    return ret;
```

Comment:
I disagree, since it exists in the scope of the main function, I added a `free` there instead:

```C
int main(int argc, char **argv)
{
    CFNetOptions opts;
    CFNetSetDefault(&opts);
    char **args = NULL;
    char *hostnames = NULL;
    int ret = CFNetParse(argc, argv,                // Inputs
                         &opts, &args, &hostnames); // Outputs
    if (ret != 0)
    {
        exit(EXIT_FAILURE);
    }
    ret = CFNetRun(&opts, args, hostnames);  // Commands return exit code
    free(hostnames);
    return ret;
}
```